db/version_set.cc

```diff
@@ -2724,90 +2726,58 @@ void VersionSet::AppendVersion(ColumnFamilyData* column_family_data,
 
   autovector<VersionEdit*> batch_edits;
-  Version* v = nullptr;
-  std::unique_ptr<BaseReferencedVersionBuilder> builder_guard(nullptr);
-
-  // process all requests in the queue
-  ManifestWriter* last_writer = &w;
-  assert(!manifest_writers_.empty());
-  assert(manifest_writers_.front() == &w);
-  if (w.edit_list.front()->IsColumnFamilyManipulation()) {
-    // no group commits for column family add or drop
-    LogAndApplyCFHelper(w.edit_list.front());
-    batch_edits.push_back(w.edit_list.front());
+  std::vector<Version*> versions;
+  std::vector<const MutableCFOptions*> mutable_cf_options_ptrs;
+  std::vector<std::unique_ptr<BaseReferencedVersionBuilder>> builder_guards;
+
+  if (first_writer.edit_list.front()->IsColumnFamilyManipulation()) {
+    // No group commits for column family add or drop
+    LogAndApplyCFHelper(first_writer.edit_list.front());
+    batch_edits.push_back(first_writer.edit_list.front());
   } else {
-    v = new Version(column_family_data, this, env_options_, mutable_cf_options,
-                    current_version_number_++);
-    builder_guard.reset(new BaseReferencedVersionBuilder(column_family_data));
-    auto* builder = builder_guard->version_builder();
-    for (const auto& writer : manifest_writers_) {
-      if (writer->edit_list.front()->IsColumnFamilyManipulation() ||
-          writer->cfd->GetID() != column_family_data->GetID()) {
+    ColumnFamilyData* cfd = nullptr;
+    Version* version = nullptr;
+    auto it = manifest_writers_.cbegin();
+    while (it != manifest_writers_.cend()) {
+      if ((*it)->edit_list.front()->IsColumnFamilyManipulation()) {
         // no group commits for column family add or drop
-        // also, group commits across column families are not supported
         break;
       }
-      last_writer = writer;
-      for (const auto& edit : writer->edit_list) {
-        LogAndApplyHelper(column_family_data, builder, v, edit, mu);
-        batch_edits.push_back(edit);
+      last_writer = *(it++);
+      if (last_writer->cfd != nullptr && last_writer->cfd->IsDropped()) {
+        continue;
+      }
+      if (cfd == nullptr || last_writer->cfd->GetID() != cfd->GetID()) {
+        cfd = last_writer->cfd;
+        version = new Version(cfd, this, env_options_,
+            last_writer->mutable_cf_options, current_version_number_++);
+        versions.push_back(version);
+        mutable_cf_options_ptrs.push_back(&last_writer->mutable_cf_options);
+        builder_guards.emplace_back(new BaseReferencedVersionBuilder(cfd));
+      }
+      auto* builder = builder_guards.back()->version_builder();
+      for (const auto& e : last_writer->edit_list) {
+        LogAndApplyHelper(cfd, builder, version, e, mu);
+        batch_edits.push_back(e);
       }
     }
-    builder->SaveTo(v->storage_info());
+    for (int i = 0; i < static_cast<int>(versions.size()); ++i) {
+      auto* builder = builder_guards[i]->version_builder();
+      builder->SaveTo(versions[i]->storage_info());
```

Comment:
I wonder what happens if the same CF appears non-consecutively in `manifest_writers_`, such that this `SaveTo` gets called twice for the same CF.

Reply:
Good catch. My assumption is that the caller of `LogAndApply` should make sure that each column family appears in the same group at most once. For atomic flush, there should be at most one version edit for each column family in the group.

Reply:
Can it happen if `ProcessManifestWrites` picks up manifest writes corresponding to multiple threads' `LogAndApply` calls?

Reply:
I think this is possible.

Reply:
@ajkr , specifically, the column family will have two new versions created for it. The `SaveTo` gets called twice, each time for a different version. What is your concern here?

Reply:
My concern was that, if there's two `Version`s for one CF, I'm not sure our calls to `SaveTo` will generate any `Version` that contains all the changes. It looks like we'll get two different `Version`s, each of which contains separate changes. Then we call `AppendVersion` on both, and only the second one becomes active.

Reply:
@ajkr Make sense. I think we can update the code so that when the same cf is seen again, the loop terminates.

Reply:
@ajkr and I discussed offline, and decided to take the following approach. In `ProcessManifestWrites`, we create one and only one version for the same column family. We can use a vector or unordered_map to store the mapping from column family id to version. Current implementation uses vector. Since this vector is usually small, we can perhaps tolerate linear scan on it. I've left a `TODO` comment there, saying `unordered_map` is an alternative.